homeassistant/components/sensor/wsdot.py

```diff
@@ -0,0 +1,145 @@
+"""
+Support for Washington State Department of Transportation (WSDOT) data.
+
+Data provided by WSDOT is documented at http://wsdot.com/traffic/api/
+"""
+import logging
+import re
+from datetime import datetime, timezone, timedelta
+
+import requests
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import (
+    CONF_API_KEY, CONF_NAME, ATTR_ATTRIBUTION,
+    )
+from homeassistant.helpers.entity import Entity
+import homeassistant.helpers.config_validation as cv
+
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_TRAVEL_TIMES = 'travel_time'
+CONF_TRAVEL_TIME_ID = 'id'
+CONF_ATTRIBUTION = "Data provided by WSDOT"
+
+# API codes for travel time details
+ATTR_ACCESS_CODE = 'AccessCode'
+ATTR_TRAVEL_TIME_ID = 'TravelTimeID'
+ATTR_CURRENT_TIME = 'CurrentTime'
+ATTR_AVG_TIME = 'AverageTime'
+ATTR_NAME = 'Name'
+ATTR_TIME_UPDATED = 'TimeUpdated'
+ATTR_DESCRIPTION = 'Description'
+
+SCAN_INTERVAL = timedelta(minutes=3)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Optional(CONF_TRAVEL_TIMES): [{
+        vol.Required(CONF_TRAVEL_TIME_ID): cv.string,
+        vol.Optional(CONF_NAME): cv.string}]
+})
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Get the WSDOT sensor."""
+    sensors = []
+    for travel_time in config.get(CONF_TRAVEL_TIMES):
+        name = (travel_time.get(CONF_NAME) or
+                travel_time.get(CONF_TRAVEL_TIME_ID))
+        sensors.append(
+            WashingtonStateTravelTimeSensor(
+                name,
+                config.get(CONF_API_KEY),
+                travel_time.get(CONF_TRAVEL_TIME_ID)))
+    add_devices(sensors)
+
+
+class WashingtonStateTransportSensor(Entity):
+    """
+    Sensor that reads the WSDOT web API.
+
+    WSDOT provides ferry schedules, toll rates, weather conditions,
+    mountain pass conditions, and more. Subclasses of this
+    can read them and make them available.
+    """
+
+    ICON = 'mdi:car'
+
+    def __init__(self, name, access_code):
+        """Initialize the sensor."""
+        self._data = {}
+        self._access_code = access_code
+        self._name = name
+        self._state = None
+        self.update()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self.ICON
+
+
+class WashingtonStateTravelTimeSensor(WashingtonStateTransportSensor):
```

Comment:
Is it to separate API code from HASS code? If so, it seems a bit superfluous to split them. There isn't a terrible amount of API logic.